uber/site_sections/budget.py

```diff
@@ -46,3 +46,157 @@ def mpoints(self, session):
         all = [(sum(mpu.amount for mpu in mpus), group, mpus)
                for group, mpus in groups.items()]
         return {'all': sorted(all, reverse=True)}
+
+    @ajax
+    def add_promo_code_words(self, session, text='', part_of_speech=0):
+        text = text.strip()
+        words = []
+        if text:
+            with Session() as session:
+                old_words = set(s for (s,) in session.query(
+                    PromoCodeWord.normalized_word).filter(
+                        PromoCodeWord.part_of_speech == part_of_speech).all())
+
+            for word in [s for s in shlex.split(text.replace(',', ' ')) if s]:
+                if PromoCodeWord.normalize_word(word) not in old_words:
+                    words.append(PromoCodeWord().apply(
+                        dict(word=word, part_of_speech=part_of_speech)))
+            words = [word.word for word in session.bulk_insert(words)]
+        return {'words': words}
+
+    @ajax
+    def delete_all_promo_code_words(self, session, part_of_speech=None):
+        query = session.query(PromoCodeWord)
+        if part_of_speech is not None:
+            query = query.filter(
+                PromoCodeWord.part_of_speech == part_of_speech)
+        result = query.delete(synchronize_session=False)
+        return {'result': result}
+
+    @ajax
+    def delete_promo_code_word(self, session, word=''):
+        result = 0
+        word = PromoCodeWord.normalize_word(word)
+        if word:
+            result = session.query(PromoCodeWord).filter(
+                PromoCodeWord.normalized_word == word).delete(
+                    synchronize_session=False)
+        return {'result': result}
+
+    def delete_promo_codes(self, session, id=None, **params):
+        query = session.query(PromoCode)
+        if id is None:
+            query = query.filter(PromoCode.uses_count == 0)
+        else:
+            ids = [s.strip() for s in id.split(',') if s.strip()]
+            query = query.filter(PromoCode.id.in_(ids))
+        result = query.delete(synchronize_session=False)
+
+        referer = cherrypy.request.headers.get('Referer', 'view_promo_codes')
```

Comment:
Ahaha, that's gotta be one of the worst and most persistent knock-on effects of someone misspelling a word in their code ever.